File: conformance/reference-js/runtime-defaults.js
'use strict';

function emptyForWire(wire) {
  if (wire === 'boolean') return false;
  if (wire === 'array') return [];
  if (wire === 'number') return null;
  return '';
}

/** ADR-0033 form field init: S0 empty → S1 defaultValue → S2 recordValues → S3 reactionWrites */
function initFormFieldValues(input) {
  const fields = Array.isArray(input.fields) ? input.fields : null;
  if (!fields) return { ok: false, code: 'INVALID_RUNTIME_DEFAULT_INPUT' };
  const values = {};
  for (const fieldDef of fields) {
    if (!fieldDef || typeof fieldDef.field !== 'string' || fieldDef.field.length === 0) {
      return { ok: false, code: 'INVALID_RUNTIME_DEFAULT_INPUT' };
    }
    const wire = fieldDef.wire || 'string';
    values[fieldDef.field] = emptyForWire(wire);
    if (Object.prototype.hasOwnProperty.call(fieldDef, 'defaultValue') && fieldDef.defaultValue !== undefined) {
      values[fieldDef.field] = fieldDef.defaultValue;
    }
  }
  if (input.recordValues && typeof input.recordValues === 'object' && !Array.isArray(input.recordValues)) {
    for (const [key, value] of Object.entries(input.recordValues)) {
      if (Object.prototype.hasOwnProperty.call(values, key)) values[key] = value;
    }
  }
  if (Array.isArray(input.reactionWrites)) {
    for (const write of input.reactionWrites) {
      if (!write || typeof write.field !== 'string') continue;
      if (Object.prototype.hasOwnProperty.call(values, write.field)) {
        values[write.field] = write.value;
      }
    }
  }
  return { ok: true, values };
}

function validateRuntimeDefaults(input) {
  if (input.kind === 'requestConfig') {
    if (input.requiresNetwork === true && (typeof input.baseURL !== 'string' || input.baseURL.trim().length === 0)) {
      return { ok: false, code: 'MISSING_BASE_URL' };
    }
    return { ok: true };
  }
  if (input.kind === 'component') {
    const installed = new Set(input.installedTypes || []);
    if (!installed.has(input.type)) return { ok: false, code: 'UNKNOWN_COMPONENT_TYPE' };
    for (const requiredProp of input.requiredProps || []) {
      if (!Object.prototype.hasOwnProperty.call(input.props || {}, requiredProp)) {
        return { ok: false, code: 'INVALID_COMPONENT', path: `props.${requiredProp}` };
      }
    }
    return { ok: true };
  }
  if (input.kind === 'defaults') {
    const value = input.value || {};
    if (input.target === 'dataRef') return { ok: true, value: { ...value, method: value.method ?? 'GET' } };
    if (input.target === 'uploadAction') {
      return {
        ok: true,
        value: {
          ...value,
          method: value.method ?? 'POST',
          retryPolicy: value.retryPolicy ?? 'never',
          fieldName: value.fieldName ?? 'file',
          multiple: value.multiple ?? false,
        },
      };
    }
  }
  if (input.kind === 'formFieldInit') {
    return initFormFieldValues(input);
  }
  return { ok: false, code: 'INVALID_RUNTIME_DEFAULT_INPUT' };
}

module.exports = { validateRuntimeDefaults };

File: conformance/reference-python/runtime_defaults.py
def _empty_for_wire(wire):
    if wire == "boolean":
        return False
    if wire == "array":
        return []
    if wire == "number":
        return None
    return ""


def _init_form_field_values(input_value):
    """ADR-0033: S0 empty → S1 defaultValue → S2 recordValues → S3 reactionWrites."""
    fields = input_value.get("fields")
    if not isinstance(fields, list):
        return {"ok": False, "code": "INVALID_RUNTIME_DEFAULT_INPUT"}
    values = {}
    for field_def in fields:
        if not isinstance(field_def, dict) or not isinstance(field_def.get("field"), str) or len(field_def["field"]) == 0:
            return {"ok": False, "code": "INVALID_RUNTIME_DEFAULT_INPUT"}
        wire = field_def.get("wire") or "string"
        name = field_def["field"]
        values[name] = _empty_for_wire(wire)
        if "defaultValue" in field_def:
            values[name] = field_def["defaultValue"]
    record_values = input_value.get("recordValues")
    if isinstance(record_values, dict):
        for key, value in record_values.items():
            if key in values:
                values[key] = value
    reaction_writes = input_value.get("reactionWrites")
    if isinstance(reaction_writes, list):
        for write in reaction_writes:
            if not isinstance(write, dict):
                continue
            field = write.get("field")
            if isinstance(field, str) and field in values:
                values[field] = write.get("value")
    return {"ok": True, "values": values}


def validate_runtime_defaults(input_value):
    kind = input_value["kind"]
    if kind == "requestConfig":
        if input_value.get("requiresNetwork") is True and not isinstance(input_value.get("baseURL"), str):
            return {"ok": False, "code": "MISSING_BASE_URL"}
        if input_value.get("requiresNetwork") is True and len(input_value["baseURL"].strip()) == 0:
            return {"ok": False, "code": "MISSING_BASE_URL"}
        return {"ok": True}
    if kind == "component":
        if input_value["type"] not in input_value.get("installedTypes", []):
            return {"ok": False, "code": "UNKNOWN_COMPONENT_TYPE"}
        props = input_value.get("props", {})
        for required_prop in input_value.get("requiredProps", []):
            if required_prop not in props:
                return {"ok": False, "code": "INVALID_COMPONENT", "path": f"props.{required_prop}"}
        return {"ok": True}
    if kind == "defaults":
        value = dict(input_value.get("value", {}))
        if input_value["target"] == "dataRef":
            value.setdefault("method", "GET")
            return {"ok": True, "value": value}
        if input_value["target"] == "uploadAction":
            value.setdefault("method", "POST")
            value.setdefault("retryPolicy", "never")
            value.setdefault("fieldName", "file")
            value.setdefault("multiple", False)
            return {"ok": True, "value": value}
    if kind == "formFieldInit":
        return _init_form_field_values(input_value)
    return {"ok": False, "code": "INVALID_RUNTIME_DEFAULT_INPUT"}
